_posts/2021-03-10-dask_memory_usage.md

```diff
@@ -0,0 +1,261 @@
+How much it increases presumably depends on the size of the files; most of them are quite small, so memory usage doesn't change much.
+**One file uses much more maximum memory than the others, 384MB of RAM; presumably it's `big.txt` which is 25MB, since the other files are all smaller than 1MB.**
+
+The mechanism used, polling peak process memory, has some limitations:
+
+* You'll notice there's a `gc.collect()` at the top of the `calculate_top_10()`; this ensures we don't count memory from previous code that hasn't been cleaned up yet.
+* There's also a `sleep()` at the bottom of `calculate_top_10()`.
+  Because polling is used, tasks that run too quickly won't get accurate information.
+* Finally, because polling is per-process, you can't run multiple threads per worker, otherwise you'll get results that combine memory usage of multiple tasks.
+
+## Interpreting the data
+
+What we've learned is that memory usage of `calculate_top_10()` grows with file size; this can let us [characterize the memory requirements for the workload](https://pythonspeed.com/articles/estimating-memory-usage/), creating a model saying "for these file sizes and this level of parallelism we will need this much RAM."
+And that can guide our hardware specifications.
+
+However, we might also notice that the memory usage seems rather high: for a 25MB file, we're using 330MB of RAM to count words.
+Thinking through how an ideal version of this code might work, we ought to be able to process the file with much less memory.
+
+**And that's another way in which `dask-memusage` can be helpful: it can point us at specific code that needs memory usage optimized, at the granularity of a task.**
+If you want to dig down further, I've built [a memory profiler called Fil](https://pythonspeed.com/fil) that is expressly designed for finding peak memory usage.
+Unlike `dask-memusage`, which can be run on production workloads, Fil slows down your execution and has other limitations I'm currently working on (it doesn't support multiple processes, as of March 2021), so for now it's better used for manual profiling.
+
+We can write a little script that only runs on `big.txt`:
+
+```python
+from pathlib import Path
+from example2 import calculate_top_10
+
+calculate_top_10(Path("files/big.txt"))
+```
+
+Run it under Fil:
+
+```shell-session
+$ pip install filprofiler
+$ fil-profile run example3.py
+```
+
+And the result shows us where the bulk of the memory is being allocated:
+
+<iframe id="peak" src="/images/dask_memusage/peak-memory.svg" width="100%" height="300" scrolling="auto" frameborder="0"></iframe>
+
+Reading in the file takes 8% of memory, but `data.split()` is responsible for 84% of memory.
+Perhaps we shouldn't be loading the whole file into memory and splitting the whole file into words, and instead we should be processing the file line by line...
```

Comment:
The elipsis ... here makes the transition to the next section feel very abrupt. After reading this sentence I expected that the logical next step was that we would change the code to process the files line by line. (It's fine if this blogpost *doesn't* do that, but you do need a sentence that finishes this idea off).